### PR TITLE
Update:  Change no-duplicate-case to comparing tokens (fixes #13485)

### LINKED ISSUE #13485: Change `no-duplicate-case` rule to comparing tokens
**What rule do you want to change?**

- `no-duplicate-case`

**Does this change cause the rule to produce more or fewer warnings?**

Produce more warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

Makes the process of detecting duplicates compare by token, like `no-dupe-else-if`.

**Please provide some example code that this change will affect:**

```js
switch (a) {
  case a.b:
  case a.b: // Already reported.
  case a.      b: // Will be reported.
  case a./**/b: // Will be reported.
}
```

**What does the rule currently do for this code?**

The comparison is text based.

**What will the rule do after it's changed?**

`a.b`, `a.    b` and `a./**/b` are reported as duplicates.

**Are you willing to submit a pull request to implement this change?**

If there is consensus, I will submit a pull request.

---

refs https://github.com/eslint/eslint/issues/13466#issuecomment-654867869

### PR DESCRIPTION
**Pull request by ota-meshi**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

fixes #13485

#### What changes did you make? (Give an overview)

This PR changes `no-duplicate-case` rule to change from detecting duplicates comparing text to detecting duplicates comparing tokens, like `no-dupe-else-if` rule.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Thanks for the PR!

I initially added "do not merge label" just to verify if this change has to be in a major release.

We're of an opinion that the current version of this rule fails to report some duplicates (those covered by this change), and that it's technically a bug.

Thus, this change can be treated as a semver-minor bug fix, so I'm removing the label now.

File: lib/rules/no-duplicate-case.js [hunk 1]
```diff
@@ -31,18 +37,31 @@ module.exports = {
...
+        /**
+         * Determines whether the two given nodes are considered to be equal.
+         * @param {ASTNode} a First node.
+         * @param {ASTNode} b Second node.
+         * @returns {boolean} `true` if the nodes are considered to be equal.
+         */
+        function equal(a, b) {
+            if (a.type !== b.type) {
+                return false;
+            }
```
Comment (anikethsaha) [inline lib/rules/no-duplicate-case.js:49]: Just a thought, 

I think this check can be moved to `astUtils.equalTokens` itself as a point of an early return.

@mdjermanovic thoughts ?

Comment (mdjermanovic) [reply lib/rules/no-duplicate-case.js:49]: That would be certainly good for optimization, but we should then probably also change the name and the description of this function (or maybe add a new function that would check the types first and then call this one). In theory, `equalTokens` check might be used to compare two different node types (e.g. a `FunctionDeclaration` and a `FunctionExpression` can have equal tokens). Maybe we could discuss this separated from this PR?

Comment (kaicataldo) [reply lib/rules/no-duplicate-case.js:49]: Agreed that we should decouple that discussion from this PR.

Comment (anikethsaha) [reply lib/rules/no-duplicate-case.js:49]: cool

**Comment by mdjermanovic** [comment]
Thanks for contributing!

**Pull request merged**